daffodil-runtime1/src/main/scala/org/apache/daffodil/dpath/NodeInfo.scala

```diff
@@ -452,21 +457,54 @@ object NodeInfo extends Enum {
       }
     }
 
+    trait PrimNonNumeric { self: AnyAtomic.Kind =>
+      protected def fromString(s: String): DataValuePrimitive
+      def fromXMLString(s: String): DataValuePrimitive = {
+        try {
+          fromString(s)
+        } catch {
+          case iae: IllegalArgumentException =>
+            throw new InvalidPrimitiveDataException("Value '%s' is not a valid %s: %s".format(s, this.globalQName, iae.getMessage))
+          case uri: URISyntaxException =>
+            throw new InvalidPrimitiveDataException("Value '%s' is not a valid %s: %s".format(s, this.globalQName, uri.getMessage))
+        }
+      }
+    }
+
     trait PrimNumeric { self: Numeric.Kind =>
-      def isValidRange(n: Number): Boolean
-      def fromNumber(n: Number): DataValueNumber
+      def isValid(n: Number): Boolean
+      protected def fromNumberNoCheck(n: Number): DataValueNumber
+      def fromNumber(n: Number): DataValueNumber = {
+        if (!isValid(n))
+          throw new InvalidPrimitiveDataException("Value '%s' is out of range for type: %s".format(n, this.globalQName))
+        val num = fromNumberNoCheck(n)
+        num
+      }
+
+      protected def fromString(s: String): DataValueNumber
+      def fromXMLString(s: String): DataValueNumber = {
+        try {
+          val num = fromString(s)
+          if (!isValid(num.getNumber))
+            throw new InvalidPrimitiveDataException("Value '%s' is out of range for type: %s".format(s, this.globalQName))
+          num
+        } catch {
+          case nfe: NumberFormatException =>
+            throw new InvalidPrimitiveDataException("Value '%s' is not a valid %s".format(s, this.globalQName))
+        }
+      }
```

Comment:
Super minor, but feels maybe a little odd to throw an exception inside a try block? Thoughs on moving it out, so something like this:
```scala
val num = try {
  fromString(s)
} catch {
   case nfd: NumberFormatException => throw new InvalidPrimitiveDataException(...)
}
if (!isValid(num.getNumber))
  throw new InvalidPrimitiveDataException(...)
num
```
Makes it more clear that the NFE catch is really just catching the fromString conversion.

Reply:
Will fix.